src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3821,6 +3821,57 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
             return this.AnonymousTypeManager.ConstructAnonymousTypeSymbol(descriptor).GetPublicSymbol();
         }
 
+        protected override IMethodSymbol CommonCreateBuiltinOperator(
+            string name,
+            ITypeSymbol returnType,
+            ITypeSymbol leftType,
+            ITypeSymbol rightType,
+            bool isChecked)
+        {
+            var csharpReturnType = returnType.EnsureCSharpSymbolOrNull(nameof(returnType));
+            var csharpLeftType = leftType.EnsureCSharpSymbolOrNull(nameof(leftType));
+            var csharpRightType = rightType.EnsureCSharpSymbolOrNull(nameof(rightType));
+
+            // caller already checked all of these were not null.
+            Debug.Assert(csharpReturnType is not null);
+            Debug.Assert(csharpLeftType is not null);
+            Debug.Assert(csharpRightType is not null);
+
+            var syntaxKind = SyntaxFacts.GetOperatorKind(name);
+            if (syntaxKind == SyntaxKind.None)
+                throw new ArgumentException($"Illegal operator name '{name}'", nameof(name));
+
+            var binaryOperatorName = OperatorFacts.BinaryOperatorNameFromSyntaxKindIfAny(syntaxKind, isChecked);
+            if (binaryOperatorName != name)
+                throw new ArgumentException($"'{name}' was not a valid binary operator name along with isChecked={isChecked}", nameof(name));
+
+            return new SynthesizedIntrinsicOperatorSymbol(csharpLeftType, name, csharpRightType, csharpReturnType, isChecked).GetPublicSymbol();
```

Comment:
I believe this code will cause us to generate operators that do not exist. For example, using `Nullable<T>` where it should be just `T`. We have a decent amount of code in the semantic model already for synthesizing intristic operators from `BoundNode`s, it feels to me like we need to extract that into a helper that can be called from here as well.

Reply:
>  it feels to me like we need to extract that into a helper that can be called from here as well.

Def would like your thoughts on how best to do that.  My understanding though from teh API notes was that it seemed like this was ok with people and that additional checks could be optional in the future.  :)

Reply:
I definitely didn't get the impression from the meeting (and it wasn't my intent to convey in the notes) that validation would optionally be added in the future. Only that we should look into it as we implemented the API. As to how to approach it: we should look at the existing places in the semantic model that create these symbols and refactor them into two components: components that extract the necessary types from the underlying bound nodes, and then the component that takes those types, does the appropriate transform (such as stripping `Nullable` where needed), and returns the resultant operator. Component 1 lives in the semantic model, 2 lives in the compilation and is called from this new api as well.

Reply:
Cool, i'll look into that!

Reply:
Talked to fred offline about this.  @AlekseyTs we'd like to know if you have some thoughts about this API.  For context, please see https://github.com/dotnet/roslyn/issues/63186

The idea here is to allow clients of the Compilation API to be able to generate 'built in' operators programatically, just like we can do for thigns like `Compilation.CreateArrayTypeSymbol` or `Compilation.CreateTupleType...` etc. etc.

These APIs have traditionally been used by the IDE so we can break symbols down to constituent parts, then reconstitute them later.  Over the years there have been some gaps here, which generally leads to certain types of symbols not working properly in some features.

One of these symbol types is the "built-in operator" IMethodSymbols.  There was no way to get back to an instance of this (without something really ugly like generating source-code and trying to bind that) :(.  We approved an API here, but have some general questions around how smart/dumb the API should be.  Currently, the API is very dumb.  It effectively doesn't do much validation (beyond name/null checks), and really just passes the inputs as-is down to the underlying SynthesizedIntrinsicOperator instance.   This works, and is totally fine, but does allow for generation of non-existent builtin operators (for example, you could generate `string DateTime.operator + (DateTime b, int? i)`).

There are two open questions here:

1. should we attempt to eb *very* strict, and only allow generation of built-in operators the language actually has?
2. when taking in inputs, should we transform them a-la: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs#L3770.  For example, removing nullability from types.

*personally*, i think we should not do anything but basic validation, and we should not transform.

The reason to not do anything but basic validation is to follow what we see elsewhere in the other creation APIs we have.  For example, while the language does not allow an array of ref-struct types, we are still allowed to create an IArrayTypeSymbol representing that.  Similarly, our other apis do not tend to transform inputs.  THey simply check that htey are respectively C# or VB symbols and pass them along.  

However, i am curious how you guys feel as general owners here with a lot of experience around the other creation APIs exposed off of the Compilation object.

Thanks!

Reply:
hi @AlekseyTs Would appreciate your thoughts here. Thank you!

Reply:
> Would appreciate your thoughts here.

I think there are the following options:
1. No support for creating symbols like that
2. Support creation
  a. No validation - feels unnecessary relaxed.
  b. Full validation is probably ideal. But I think it will be hard to implement, because it looks like the existing code and logic is not structured in a way that allows easy reuse for validation purposes. For example, operator overload resolution looks for a suitable intrinsic operator for two expressions (not types), and the one it finds might not match the operand types exactly. Plus enum operators, UTF8 concatenation, etc.
  c. Limited validation - perhaps we can come up with not extremely complicated validation heuristic that will detect obvious errors like: use of ```Nullable<T>```, different parameter types for intrinsic types, arbitrary types (not enums, delegates, intrinsic, etc) .

Reply:
> Limited validation - perhaps we can come up with not extremely complicated validation heuristic that will detect obvious errors like: use of Nullable<T>, different parameter types for intrinsic types, arbitrary types (not enums, delegates, intrinsic, etc) .

Let me look into this to see what i can do.  Thanks!

Reply:
I think you do support synthesizing ops for enums btw:

![image](https://user-images.githubusercontent.com/4564579/186268800-03ad3ba8-2160-4008-b9b1-ec8e4470f195.png)

Or did you mean something else? :)

Reply:
> I think you do support synthesizing ops for enums btw

I think I mentioned them: "Plus **enum** operators, UTF8 concatenation, etc."

Reply:
yeah, i read 'not enums' as 'disallow enums'.  As opposed to "don't allow aribtrary types, but *do* allow enums" :)

Reply:
interesting thing i've found.  'dynamic' really gums up the works.  Compiler will happily synthesize intrinsic operators between it and virtually any other type i've checked.  i either have to special case dynamic (and basically then have no checks) or allow everything.